bfvmm/include/vcpu/arch/intel_x64/efi_vcpu.h

```diff
@@ -0,0 +1,152 @@
+
+#include <eapis/vcpu/arch/intel_x64/vcpu.h>
+#include "../../../hve/arch/intel_x64/hve.h"
+#include "../../../hve/arch/intel_x64/vic.h"
+
+using namespace eapis::intel_x64;
+
+namespace efi
+{
+
+/// EFI vCPU
+///
+/// This class provides a vCPU for use when booting directly from EFI
+///
+class vcpu : public eapis::intel_x64::vcpu
+{
+public:
+
+    /// Constructor
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    /// @param id the id of this vcpu
+    ///
+    vcpu(vcpuid::type id) :
+        eapis::intel_x64::vcpu{id},
+        m_hve{std::make_unique<eapis::intel_x64::hve>(exit_handler(), vmcs())}
+    {
+        hve()->enable_wrcr0_exiting(
+            0xFFFFFFFFFFFFFFFF, ::intel_x64::vmcs::guest_cr0::get()
+        );
+
+        hve()->add_wrcr0_handler(
+            control_register::handler_delegate_t::create<vcpu, &vcpu::handle_wrcr0>(this)
+        );
+
+        hve()->enable_wrcr4_exiting(
+            ::intel_x64::cr4::vmx_enable_bit::mask, ::intel_x64::vmcs::guest_cr4::get()
+        );
+
+        hve()->add_wrcr4_handler(
+            control_register::handler_delegate_t::create<vcpu, &vcpu::handle_wrcr4>(this)
+        );
+
+        exit_handler()->add_handler(
+            ::intel_x64::vmcs::exit_reason::basic_exit_reason::cpuid,
+            ::handler_delegate_t::create<vcpu, &vcpu::handle_cpuid>(this)
+            );
+
+        exit_handler()->add_handler(
+            ::intel_x64::vmcs::exit_reason::basic_exit_reason::rdmsr,
+            ::handler_delegate_t::create<vcpu, &vcpu::handle_rdmsr>(this)
+            );
+
+        exit_handler()->add_handler(
+            ::intel_x64::vmcs::exit_reason::basic_exit_reason::wrmsr,
+            ::handler_delegate_t::create<vcpu, &vcpu::handle_wrmsr>(this)
+            );
+
+        exit_handler()->add_handler(
+            ::intel_x64::vmcs::exit_reason::basic_exit_reason::init_signal,
+            ::handler_delegate_t::create<vcpu, &vcpu::handle_init_signal>(this)
+            );
+
+        exit_handler()->add_handler(
+            ::intel_x64::vmcs::exit_reason::basic_exit_reason::sipi,
+            ::handler_delegate_t::create<vcpu, &vcpu::handle_sipi>(this)
+            );
+
+        exit_handler()->add_handler(
+            ::intel_x64::vmcs::exit_reason::basic_exit_reason::ept_violation,
+            ::handler_delegate_t::create<vcpu, &vcpu::handle_ept_violation>(this)
+            );
+
+        exit_handler()->add_handler(
+            ::intel_x64::vmcs::exit_reason::basic_exit_reason::vmcall,
+            ::handler_delegate_t::create<vcpu, &vcpu::handle_vmcall>(this)
+            );
+    }
+
+    /// Destructor
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    ~vcpu() = default;
+
+    /// Get HVE (hardware virtualization extensions)
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    /// @return Returns the hve object stored in this vCPU
+    ///
+    gsl::not_null<eapis::intel_x64::hve *> hve()
```

Comment:
you should be able to get the hve and vic from the eapis, they shouldn't be redefined here